File: app/src/main/java/com/kozhevin/rootchecks/ui/MainActivity.java
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017  Dmitrii Kozhevin <kozhevin.dima@gmail.com>
 *
 *  Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 *  and associated documentation files (the “Software”), to deal in the Software without
 *  restriction, including without limitation the rights to use, copy, modify, merge, publish,
 *  distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the
 *  Software is furnished to do so, subject to the following conditions:
 *
 *  The above copyright notice and this permission notice shall be included in all copies or
 *  substantial portions of the Software.
 *
 *  THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 *  BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 *  DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 *  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

package com.kozhevin.rootchecks.ui;


import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;

import com.kozhevin.rootchecks.R;
import com.kozhevin.rootchecks.data.CheckInfo;
import com.kozhevin.rootchecks.data.TotalResult;
import com.kozhevin.rootchecks.util.CheckTask;
import com.kozhevin.rootchecks.util.IChecksResultListener;

import java.util.ArrayList;

import static com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_CHECKED_ERROR;
import static com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_CHECKED_ROOT_DETECTED;
import static com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_CHECKED_ROOT_NOT_DETECTED;
import static com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_STILL_GOING;
import static com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_UNCHECKED;
import static com.kozhevin.rootchecks.constant.GeneralConst.GITHUB;

public class MainActivity extends AppCompatActivity implements IChecksResultListener {

    private CheckTask mTask;
    private RecyclerView mList;
    private ImageView mResultState;
    private FloatingActionButton mFapMe; // I'm so sorry ;)
    private ProgressBar mProgress;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle("");
        setSupportActionBar(toolbar);
        mProgress = findViewById(R.id.result_progress);
        mProgress.setVisibility(View.GONE);
        mResultState = findViewById(R.id.result_image_view);
        mList = findViewById(R.id.list);
        mList.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        mList.setAdapter(new ResultAdapter());
        mFapMe = findViewById(R.id.fab);
        mFapMe.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (mTask != null) {
                    mTask.cancel(false);
                }
                mTask = new CheckTask(MainActivity.this, false);
                mTask.execute((Void[]) null);
            }
        });
        mTask = new CheckTask(MainActivity.this, true);
        mTask.execute((Void[]) null);

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_github) {
            Intent i = new Intent(Intent.ACTION_VIEW);
            i.setData(Uri.parse(GITHUB));
            PackageManager manager = getPackageManager();
            if (i.resolveActivity(manager) != null) {
                startActivity(i);
            }
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onProcessStarted() {
        mResultState.setVisibility(View.INVISIBLE);
        mFapMe.hide();
        mProgress.setVisibility(View.VISIBLE);
    }

    @Override
    public void onUpdateResult(TotalResult result) {
        if (result != null && mList.getAdapter() != null) {
            ((ResultAdapter) mList.getAdapter()).setData((ArrayList<CheckInfo>) result.getList());
        }
    }

    @Override
    public void onProcessFinished(TotalResult result) {
        if (mTask != null) {
            mTask = null;
        }
        mProgress.setVisibility(View.GONE);
        mFapMe.show();
        if (result != null) {
            ((ResultAdapter) mList.getAdapter()).setData((ArrayList<CheckInfo>) result.getList());

            switch (result.getCheckState()) {

                case CH_STATE_CHECKED_ROOT_DETECTED:
                    mResultState.setVisibility(View.VISIBLE);
                    mResultState.setImageResource(R.drawable.ic_rooted);
                    break;

                case CH_STATE_CHECKED_ROOT_NOT_DETECTED:
                    mResultState.setVisibility(View.VISIBLE);
                    mResultState.setImageResource(R.drawable.ic_notrooted);
                    break;

                case CH_STATE_STILL_GOING:
                    mResultState.setVisibility(View.INVISIBLE);
                    break;

                case CH_STATE_UNCHECKED:
                    mResultState.setVisibility(View.INVISIBLE);
                    break;

                case CH_STATE_CHECKED_ERROR:
                    mResultState.setVisibility(View.INVISIBLE);
                    break;

                default:
                    throw new IllegalStateException("Unknown state of the result");
            }
        }

    }

}

File: appKotlin/src/main/java/com/kozhevin/rootchecks/ui/MainActivity.kt
package com.kozhevin.rootchecks.ui

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.ProgressBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.kozhevin.rootchecks.R
import com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_CHECKED_ERROR
import com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_CHECKED_ROOT_DETECTED
import com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_CHECKED_ROOT_NOT_DETECTED
import com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_STILL_GOING
import com.kozhevin.rootchecks.constant.GeneralConst.CH_STATE_UNCHECKED
import com.kozhevin.rootchecks.constant.GeneralConst.GITHUB
import com.kozhevin.rootchecks.data.CheckInfo
import com.kozhevin.rootchecks.data.TotalResult
import com.kozhevin.rootchecks.util.CheckTask
import com.kozhevin.rootchecks.util.IChecksResultListener
import java.util.*

class MainActivity : AppCompatActivity(), IChecksResultListener {
    private var mTask: CheckTask? = null
    private var mList: RecyclerView? = null
    private var mResultState: ImageView? = null
    private var mFab: FloatingActionButton? = null
    private var mProgress: ProgressBar? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        toolbar.title = ""
        setSupportActionBar(toolbar)
        mProgress = findViewById(R.id.result_progress)
        mProgress!!.visibility = View.GONE
        mResultState = findViewById(R.id.result_image_view)
        mList = findViewById(R.id.list)
        mList!!.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        mList!!.adapter = ResultAdapter()
        mFab = findViewById(R.id.fab)
        mFab!!.setOnClickListener {
            mTask?.cancel(false)
            mTask = CheckTask(this@MainActivity, false)
            mTask!!.execute()
        }
        mTask = CheckTask(this@MainActivity, true)
        mTask!!.execute()

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = if (item.itemId == R.id.action_github) {
        val i = Intent(Intent.ACTION_VIEW, Uri.parse(GITHUB))
        if (i.resolveActivity(packageManager) != null) {
            startActivity(i)
        }
        true
    } else {
        super.onOptionsItemSelected(item)
    }

    override fun onProcessStarted() {
        mResultState!!.visibility = View.INVISIBLE
        mFab!!.hide()
        mProgress!!.visibility = View.VISIBLE
    }

    override fun onUpdateResult(result: TotalResult) {
        if (mList!!.adapter != null) {
            (mList!!.adapter as ResultAdapter).setData(result.list as ArrayList<CheckInfo>)
        }
    }

    override fun onProcessFinished(result: TotalResult) {
        if (mTask != null) {
            mTask = null
        }
        mProgress!!.visibility = View.GONE
        mFab!!.show()
        (mList!!.adapter as ResultAdapter).setData(result.list as ArrayList<CheckInfo>)
        when (result.checkState) {
            CH_STATE_CHECKED_ROOT_DETECTED -> {
                mResultState!!.visibility = View.VISIBLE
                mResultState!!.setImageResource(R.drawable.ic_rooted)
            }
            CH_STATE_CHECKED_ROOT_NOT_DETECTED -> {
                mResultState!!.visibility = View.VISIBLE
                mResultState!!.setImageResource(R.drawable.ic_notrooted)
            }
            CH_STATE_STILL_GOING -> mResultState!!.visibility = View.INVISIBLE
            CH_STATE_UNCHECKED -> mResultState!!.visibility = View.INVISIBLE
            CH_STATE_CHECKED_ERROR -> mResultState!!.visibility = View.INVISIBLE
            else -> throw IllegalStateException("Unknown state of the result")
        }
    }
}
